datasets/sofc_materials_articles/sofc_materials_articles.py

```diff
@@ -0,0 +1,470 @@
+
+        text_base_path = os.path.join(data_dir, "texts")
+
+        text_files_available = [
+            os.path.split(i.rstrip(".txt"))[-1] for i in glob.glob(os.path.join(text_base_path, "*.txt"))
+        ]
+
+        metadata = metadata[metadata["name"].map(lambda x: x in text_files_available)]
+
+        names = {}
+        splits = ["train", "test", "dev"]
+        for split in splits:
+            names[split] = metadata[metadata["set"] == split]["name"].tolist()
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "names": names["train"],
+                    "data_dir": data_dir,
+                    "split": "train",
+                },
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.TEST,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={"names": names["test"], "data_dir": data_dir, "split": "test"},
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.VALIDATION,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "names": names["dev"],
+                    "data_dir": data_dir,
+                    "split": "validation",
+                },
+            ),
+        ]
+
+    def _generate_examples(self, names, data_dir, split):
+        """ Yields examples. """
+        # The dataset consists of the original article text as well as annotations
+        textfile_base_path = os.path.join(data_dir, "texts")
+        annotations_base_path = os.path.join(data_dir, "annotations")
+
+        # The annotations are mostly references to offsets in the source text
+        # with corresponding labels, so we'll refer to them as `meta`
+        sentence_meta_base_path = os.path.join(annotations_base_path, "sentences")
+        tokens_meta_base_path = os.path.join(annotations_base_path, "tokens")
+        ets_meta_base_path = os.path.join(annotations_base_path, "entity_types_and_slots")
+        frame_meta_base_path = os.path.join(annotations_base_path, "frames")
+
+        # Define the headers for the sentence and token and entity metadata
+        sentence_meta_header = ["sentence_id", "label", "begin_char_offset", "end_char_offset"]
+        tokens_meta_header = ["sentence_id", "token_id", "begin_char_offset", "end_char_offset"]
+        ets_meta_header = [
+            "sentence_id",
+            "token_id",
+            "begin_char_offset",
+            "end_char_offset",
+            "entity_label",
+            "slot_label",
+        ]
+
+        # Start the processing loop
+        # For each text file, we'll load all of the
+        # associated annotation files
+        for id_, name in enumerate(names):
```

Comment:
```suggestion
        for id_, name in enumerate(sorted(names)):
```
we need to make the order deterministic here